### DISCUSSION TITLE
Autocompletion and type checking

### DISCUSSION THREAD

**Discussion by reignmaker** [PRQL/prql#4156] [Q&A]
Hey there!

I was reading the docs and didn't find any mention of whether PRQL provides autocompletion on objects (relations) and type checking before (or after?) compilation. Does it have support of LSP? How DX in general feels like?

**Comment by vanillajonathan** [comment]
No, PRQL does not provide any autocompletion for objects (relations) and it does yet have LSP but it is on the [roadmap](https://prql-lang.org/roadmap/). Do you have any experience with LSP, maybe you could help out?

PRQL is just a transpiler that outputs SQL so it not possible to do type checking after, so currently the types are just type hints. @aljazerzen is working on a query engine called Lutra in PR #4134 maybe it could do type checking.

The DX depends on how you use it I guess, if you use the VS Code extension, the JetBrains IDE plugin, the online playground, or the `prqlc` CLI, or any of the many language bindings.
